Confusing signer node exception message on incorrect Ethereum address

Hi.

I've attempted to publish a Demand with wrong Validator address:
```console
rostopic pub -1 /liability/infochan/eth/signing/demand robonomics_msgs/Demand '{model: {multihash: "QmW3dTa1QZxnZzpF9TDuHKN7GDKaJDVoQys4u38xmdAGHF"}, objective: {multihash: "QmUA5ZHHqFZyR1T1KxxheyrUzEEymfKb3goPZt1jLDNEWL"}, token: {address: "0x368C35597A9068280028c424122DE9129fAF8689"}, cost: {uint256: "1"}, lighthouse: {address: "0x202a09A451DE674d2d65Bf1C90968a8d8F72cf7b"}, validator: {address: "0x0x368C35597A9068280028c424122DE9129fAF8689"}, validatorFee: {uint256: "0"}, deadline: {uint256: "210000"}, sender: {address: ""}, signature: []}'
```

As you can see, validator address contains double _0x_ prefix. There is a _Signer_ node log below where you can find it raise a web3 exception:

```console
...
web3.exceptions.InvalidAddress: Could not look up name '0x0x368C35597A9068280028c424122DE9129fAF8689' because no web3 connection available
```

Exception name _InvalidAddress_ gives a tip, but message "no web3 connection" makes confuse.

<details>
  <summary>liability-infochan-eth-signer-4.log</summary>

```console
[rospy.client][INFO] 2019-06-12 13:34:28,654: init_node, name[/liability/infochan/eth/signer], pid[1078]
[xmlrpc][INFO] 2019-06-12 13:34:28,661: XML-RPC server binding to 0.0.0.0:0
[xmlrpc][INFO] 2019-06-12 13:34:28,661: Started XML-RPC server [http://iotipciAira:33477/]
[rospy.impl.masterslave][INFO] 2019-06-12 13:34:28,662: _ready: http://iotipciAira:33477/
[rospy.init][INFO] 2019-06-12 13:34:28,671: ROS Slave URI: [http://iotipciAira:33477/]
[rospy.registration][INFO] 2019-06-12 13:34:28,672: Registering with master node http://localhost:11311
[xmlrpc][INFO] 2019-06-12 13:34:28,679: xml rpc node: starting XML-RPC server
[rospy.init][INFO] 2019-06-12 13:34:28,771: registered with master
[rospy.rosout][INFO] 2019-06-12 13:34:28,772: initializing /rosout core topic
[rospy.rosout][INFO] 2019-06-12 13:34:28,774: connected to core topic /rosout
[rospy.simtime][INFO] 2019-06-12 13:34:28,776: /use_sim_time is not set, will not subscribe to simulated time [/clock] topic
[rospy.internal][INFO] 2019-06-12 13:34:29,005: topic[/rosout] adding connection to [/rosout], count 0
[rospy.internal][INFO] 2019-06-12 13:34:33,617: topic[/liability/infochan/eth/sending/demand] adding connection to [/liability/infochan/ipfs_channel], count 0
[rospy.internal][INFO] 2019-06-12 13:34:33,626: topic[/liability/infochan/eth/sending/offer] adding connection to [/liability/infochan/ipfs_channel], count 0
[rospy.internal][INFO] 2019-06-12 13:34:33,631: topic[/liability/infochan/eth/sending/result] adding connection to [/liability/infochan/ipfs_channel], count 0
[rospy.internal][INFO] 2019-06-12 13:34:33,649: topic[/liability/infochan/eth/signing/result] adding connection to [http://iotipciAira:46633/], count 0
[rospy.internal][INFO] 2019-06-12 13:49:01,075: topic[/liability/infochan/eth/signing/demand] adding connection to [http://iotipciAira:37661/], count 0
[rospy.internal][INFO] 2019-06-12 13:49:01,081: topic[/liability/infochan/eth/signing/offer] adding connection to [http://iotipciAira:37661/], count 0

...

[rospy.internal][INFO] 2019-06-13 11:19:40,831: topic[/liability/infochan/eth/signing/demand] removing connection to http://iotipciAira:46481/
[rospy.internal][INFO] 2019-06-13 11:19:44,556: topic[/liability/infochan/eth/signing/demand] adding connection to [http://iotipciAira:40641/], count 0
[rosout][ERROR] 2019-06-13 11:19:44,572: bad callback: <function Signer.__init__.<locals>.sign_demand at 0x7f4f66c9fd90>
Traceback (most recent call last):
  File "/nix/store/z5qzs91jqskq6kgyad93q5cid6z0vvhv-rospy-1.14.4/lib/python3.7/site-packages/rospy/topics.py", line 750, in _invoke_callback
    cb(msg)
  File "/nix/store/psxml9bk7d4g61g3lvxy8fm7gnib5bg0-robonomics_comm-1.0.0/lib/python3.7/site-packages/ethereum_common/signer.py", line 83, in sign_demand
    message_hash = robonomicsMessageUtils.demand_hash(msg, current_nonce)
  File "/nix/store/psxml9bk7d4g61g3lvxy8fm7gnib5bg0-robonomics_comm-1.0.0/lib/python3.7/site-packages/robonomics_msgs/robonomicsMessageUtils.py", line 32, in demand_hash
    msg.sender.address])
  File "/nix/store/d2b47n76nwwa4c9a1vnrd780rmcxpnxp-python3.7-web3-4.8.3/lib/python3.7/site-packages/web3/utils/decorators.py", line 16, in _wrapper
    return self.method(objtype, *args, **kwargs)
  File "/nix/store/d2b47n76nwwa4c9a1vnrd780rmcxpnxp-python3.7-web3-4.8.3/lib/python3.7/site-packages/web3/main.py", line 167, in soliditySha3
    normalized_values = map_abi_data([abi_ens_resolver(w3)], abi_types, values)
  File "cytoolz/functoolz.pyx", line 232, in cytoolz.functoolz.curry.__call__
  File "/nix/store/d2b47n76nwwa4c9a1vnrd780rmcxpnxp-python3.7-web3-4.8.3/lib/python3.7/site-packages/web3/utils/abi.py", line 442, in map_abi_data
    return pipe(data, *pipeline)
  File "cytoolz/functoolz.pyx", line 589, in cytoolz.functoolz.pipe
  File "cytoolz/functoolz.pyx", line 565, in cytoolz.functoolz.c_pipe
  File "cytoolz/functoolz.pyx", line 232, in cytoolz.functoolz.curry.__call__
  File "/nix/store/d2b47n76nwwa4c9a1vnrd780rmcxpnxp-python3.7-web3-4.8.3/lib/python3.7/site-packages/web3/utils/abi.py", line 474, in data_tree_map
    return recursive_map(map_to_typed_data, data_tree)
  File "/nix/store/d2b47n76nwwa4c9a1vnrd780rmcxpnxp-python3.7-web3-4.8.3/lib/python3.7/site-packages/web3/utils/decorators.py", line 35, in wrapped
    wrapped_val = to_wrap(*args)
  File "/nix/store/d2b47n76nwwa4c9a1vnrd780rmcxpnxp-python3.7-web3-4.8.3/lib/python3.7/site-packages/web3/utils/formatters.py", line 115, in recursive_map
    items_mapped = map_collection(recurse, data)
  File "/nix/store/d2b47n76nwwa4c9a1vnrd780rmcxpnxp-python3.7-web3-4.8.3/lib/python3.7/site-packages/web3/utils/formatters.py", line 102, in map_collection
    return datatype(map(func, collection))
  File "/nix/store/d2b47n76nwwa4c9a1vnrd780rmcxpnxp-python3.7-web3-4.8.3/lib/python3.7/site-packages/web3/utils/formatters.py", line 114, in recurse
    return recursive_map(func, item)
  File "/nix/store/d2b47n76nwwa4c9a1vnrd780rmcxpnxp-python3.7-web3-4.8.3/lib/python3.7/site-packages/web3/utils/decorators.py", line 35, in wrapped
    wrapped_val = to_wrap(*args)
  File "/nix/store/d2b47n76nwwa4c9a1vnrd780rmcxpnxp-python3.7-web3-4.8.3/lib/python3.7/site-packages/web3/utils/formatters.py", line 116, in recursive_map
    return func(items_mapped)
  File "/nix/store/d2b47n76nwwa4c9a1vnrd780rmcxpnxp-python3.7-web3-4.8.3/lib/python3.7/site-packages/web3/utils/abi.py", line 471, in map_to_typed_data
    return ABITypedData(func(*elements))
  File "cytoolz/functoolz.pyx", line 232, in cytoolz.functoolz.curry.__call__
  File "/nix/store/d2b47n76nwwa4c9a1vnrd780rmcxpnxp-python3.7-web3-4.8.3/lib/python3.7/site-packages/web3/utils/normalizers.py", line 140, in abi_ens_resolver
    " connection available" % (val)
web3.exceptions.InvalidAddress: Could not look up name '0x0x368C35597A9068280028c424122DE9129fAF8689' because no web3 connection available

[rospy.internal][ERROR] 2019-06-13 11:19:47,558: Unable to initiate TCP/IP socket to iotipciAira:34629 (http://iotipciAira:40641/): Traceback (most recent call last):
  File "/nix/store/z5qzs91jqskq6kgyad93q5cid6z0vvhv-rospy-1.14.4/lib/python3.7/site-packages/rospy/impl/tcpros_base.py", line 558, in connect
    self.read_header()
  File "/nix/store/z5qzs91jqskq6kgyad93q5cid6z0vvhv-rospy-1.14.4/lib/python3.7/site-packages/rospy/impl/tcpros_base.py", line 653, in read_header
    self._validate_header(read_ros_handshake_header(sock, self.read_buff, self.protocol.buff_size))
  File "/nix/store/z5qzs91jqskq6kgyad93q5cid6z0vvhv-rospy-1.14.4/lib/python3.7/site-packages/rospy/impl/tcpros_base.py", line 596, in _validate_header
    raise TransportInitError("remote error reported: %s"%header['error'])
rospy.exceptions.TransportInitError: remote error reported: node shutting down

[rospy.internal][WARNING] 2019-06-13 11:19:48,560: Unknown error initiating TCP/IP socket to iotipciAira:34629 (http://iotipciAira:40641/): Traceback (most recent call last):
  File "/nix/store/z5qzs91jqskq6kgyad93q5cid6z0vvhv-rospy-1.14.4/lib/python3.7/site-packages/rospy/impl/tcpros_base.py", line 556, in connect
    self.socket.connect((dest_addr, dest_port))
ConnectionRefusedError: [Errno 111] Connection refused

[rospy.internal][INFO] 2019-06-13 11:19:48,560: topic[/liability/infochan/eth/signing/demand] removing connection to http://iotipciAira:40641/
```
</details>

--- a/ethereum_common/src/ethereum_common/signer.py
+++ b/ethereum_common/src/ethereum_common/signer.py
@@ -9,7 +9,7 @@ from web3 import Web3, HTTPProvider
 from web3.exceptions import StaleBlockchain
 from threading import Lock
 from ens import ENS
-from robonomics_msgs import robonomicsMessageUtils
+from robonomics_msgs import robonomicsMessageUtils, messageValidator
 import rospy
 import binascii
 import json
@@ -84,35 +84,46 @@ class Signer:
 
         def sign_demand(msg):
             msg.sender = Address(self.__account.address)
-            msg = robonomicsMessageUtils.convert_msg_ens_names_to_addresses(msg, web3=self.web3)
-            current_nonce = get_nonce_by_address(msg.sender)
-            message_hash = robonomicsMessageUtils.demand_hash(msg, current_nonce)
-            signed_hash = self.__account.signHash(defunct_hash_message(message_hash))
-            msg.signature = signed_hash.signature
-            msg.nonce = UInt256(uint256=str(current_nonce.uint256))
-            rospy.loginfo('askhash: %s signature: %s', binascii.hexlify(message_hash), binascii.hexlify(msg.signature))
-            self.signed_demand.publish(msg)
+
+            if messageValidator.isDemandFieldsCorrect(msg):
+                msg = robonomicsMessageUtils.convert_msg_ens_names_to_addresses(msg, web3=self.web3)
+                current_nonce = get_nonce_by_address(msg.sender)
+                message_hash = robonomicsMessageUtils.demand_hash(msg, current_nonce)
+                signed_hash = self.__account.signHash(defunct_hash_message(message_hash))
+                msg.signature = signed_hash.signature
+                msg.nonce = UInt256(uint256=str(current_nonce.uint256))
+                rospy.loginfo('askhash: %s signature: %s', binascii.hexlify(message_hash), binascii.hexlify(msg.signature))
+                self.signed_demand.publish(msg)
+            else:
+                rospy.logerr("Signing demand error: msg %s is not valid Demand message", msg)
         rospy.Subscriber('signing/demand', Demand, sign_demand)
 
         def sign_offer(msg):
             msg.sender = Address(self.__account.address)
-            msg = robonomicsMessageUtils.convert_msg_ens_names_to_addresses(msg, web3=self.web3)
-            current_nonce = get_nonce_by_address(msg.sender)
-            message_hash = robonomicsMessageUtils.offer_hash(msg, current_nonce)
-            signed_hash = self.__account.signHash(defunct_hash_message(message_hash))
-            msg.signature = signed_hash.signature
-            msg.nonce = UInt256(uint256=str(current_nonce.uint256))
-            rospy.loginfo('bidhash: %s signature: %s', binascii.hexlify(message_hash), binascii.hexlify(msg.signature))
-            self.signed_offer.publish(msg)
+
+            if messageValidator.isOfferFieldsCorrect(msg):
+                msg = robonomicsMessageUtils.convert_msg_ens_names_to_addresses(msg, web3=self.web3)
+                current_nonce = get_nonce_by_address(msg.sender)
+                message_hash = robonomicsMessageUtils.offer_hash(msg, current_nonce)
+                signed_hash = self.__account.signHash(defunct_hash_message(message_hash))
+                msg.signature = signed_hash.signature
+                msg.nonce = UInt256(uint256=str(current_nonce.uint256))
+                rospy.loginfo('bidhash: %s signature: %s', binascii.hexlify(message_hash), binascii.hexlify(msg.signature))
+                self.signed_offer.publish(msg)
+            else:
+                rospy.logerr("Signing offer error: msg %s is not valid Offer message", msg)
         rospy.Subscriber('signing/offer', Offer, sign_offer)
 
         def sign_result(msg):
-            msg = robonomicsMessageUtils.convert_msg_ens_names_to_addresses(msg, web3=self.web3)
-            message_hash = robonomicsMessageUtils.result_hash(msg)
-            signed_hash = self.__account.signHash(defunct_hash_message(message_hash))
-            msg.signature = signed_hash.signature
-            rospy.loginfo('reshash: %s signature: %s', binascii.hexlify(message_hash), binascii.hexlify(msg.signature))
-            self.signed_result.publish(msg)
+            if messageValidator.isResultFieldsCorrect(msg):
+                msg = robonomicsMessageUtils.convert_msg_ens_names_to_addresses(msg, web3=self.web3)
+                message_hash = robonomicsMessageUtils.result_hash(msg)
+                signed_hash = self.__account.signHash(defunct_hash_message(message_hash))
+                msg.signature = signed_hash.signature
+                rospy.loginfo('reshash: %s signature: %s', binascii.hexlify(message_hash), binascii.hexlify(msg.signature))
+                self.signed_result.publish(msg)
+            else:
+                rospy.logerr("Signing result error: msg %s is not valid Result message")
         rospy.Subscriber('signing/result', Result, sign_result)
 
     def spin(self):

--- a/ipfs_common/src/ipfs_common/ipfs_channel.py
+++ b/ipfs_common/src/ipfs_common/ipfs_channel.py
@@ -5,53 +5,12 @@
 
 from robonomics_msgs.msg import Demand, Offer, Result
 from robonomics_msgs.messageValidator import convertMessage
-from binascii import hexlify
+from robonomics_msgs.robonomicsMessageUtils import offer2dict, demand2dict, res2dict
 from .pubsub import publish, subscribe
 from urllib.parse import urlparse
 from threading import Thread
 import rospy
 import ipfsapi
-
-
-def bid2dict(b):
-    return {
-        'model': b.model.multihash,
-        'objective': b.objective.multihash,
-        'token': b.token.address,
-        'cost': int(b.cost.uint256),
-        'validator': b.validator.address,
-        'lighthouse': b.lighthouse.address,
-        'lighthouseFee': int(b.lighthouseFee.uint256),
-        'deadline': int(b.deadline.uint256),
-        'sender': b.sender.address,
-        'nonce': int(b.nonce.uint256),
-        'signature': hexlify(b.signature).decode('utf-8')
-    }
-
-
-def ask2dict(a):
-    return {
-        'model': a.model.multihash,
-        'objective': a.objective.multihash,
-        'token': a.token.address,
-        'cost': int(a.cost.uint256),
-        'lighthouse': a.lighthouse.address,
-        'validator': a.validator.address,
-        'validatorFee': int(a.validatorFee.uint256),
-        'deadline': int(a.deadline.uint256),
-        'sender': a.sender.address,
-        'nonce': int(a.nonce.uint256),
-        'signature': hexlify(a.signature).decode('utf-8')
-    }
-
-
-def res2dict(r):
-    return {
-        'liability': r.liability.address,
-        'result': r.result.multihash,
-        'success': r.success,
-        'signature': hexlify(r.signature).decode('utf-8')
-    }
 
 
 class IPFSChannel:
@@ -68,8 +27,8 @@ class IPFSChannel:
         self.incoming_demand = rospy.Publisher('incoming/demand', Demand, queue_size=10)
         self.incoming_result = rospy.Publisher('incoming/result', Result, queue_size=10)
 
-        rospy.Subscriber('eth/sending/offer',  Offer,  lambda m: publish(self.ipfs_client, self.lighthouse, bid2dict(m)))
-        rospy.Subscriber('eth/sending/demand', Demand, lambda m: publish(self.ipfs_client, self.lighthouse, ask2dict(m)))
+        rospy.Subscriber('eth/sending/offer',  Offer,  lambda m: publish(self.ipfs_client, self.lighthouse, offer2dict(m)))
+        rospy.Subscriber('eth/sending/demand', Demand, lambda m: publish(self.ipfs_client, self.lighthouse, demand2dict(m)))
         rospy.Subscriber('eth/sending/result', Result, lambda m: publish(self.ipfs_client, self.lighthouse, res2dict(m)))
 
     def spin(self):

--- a/robonomics_liability/tests/test_executor.py
+++ b/robonomics_liability/tests/test_executor.py
@@ -3,7 +3,7 @@
 import unittest, rostest, os, time, rospy, rosbag
 
 from robonomics_msgs.msg import Result, Offer, Demand
-from robonomics_msgs import messageValidator
+from robonomics_msgs import robonomicsMessageUtils
 from robonomics_liability.msg import Liability
 from robonomics_liability.srv import FinishLiability, StartLiability
 from urllib.parse import urlparse
@@ -136,7 +136,7 @@ class TestExecutor(unittest.TestCase):
             "nonce": 0,
             "signature": ""
         }
-        return messageValidator.dict2bid(bidDict)
+        return robonomicsMessageUtils.dict2offer(bidDict)
 
     def get_test_ask(self):
         askDict = {
@@ -152,7 +152,7 @@ class TestExecutor(unittest.TestCase):
             "nonce": 0,
             "signature": ""
         }
-        return messageValidator.dict2ask(askDict)
+        return robonomicsMessageUtils.dict2demand(askDict)
 
 
 if __name__ == '__main__':

--- a/robonomics_liability/tests/test_infochan.py
+++ b/robonomics_liability/tests/test_infochan.py
@@ -3,6 +3,7 @@
 import rostest, sys, unittest, rospy, time
 from ipfs_common import ipfs_channel
 from robonomics_msgs.msg import Demand, Offer, Result
+from robonomics_msgs.robonomicsMessageUtils import offer2dict, demand2dict, res2dict
 import testMessages
 
 PKG = 'robonomics_liability'
@@ -26,14 +27,14 @@ class TestInfochan(unittest.TestCase):
         self.infochan_Res_subscriber_topic = rospy.Publisher('/liability/infochan/eth/sending/result', Result, queue_size=10)
 
 
-    def test_ask2dict(self):
-        self.assertEqual(testMessages.validAskDict, ipfs_channel.ask2dict(testMessages.getValidAsk()))
+    def test_demand2dict(self):
+        self.assertEqual(testMessages.validAskDict, demand2dict(testMessages.getValidAsk()))
 
-    def test_bid2dict(self):
-        self.assertEqual(testMessages.validBidDict, ipfs_channel.bid2dict(testMessages.getValidBid()))
+    def test_offer2dict(self):
+        self.assertEqual(testMessages.validBidDict, offer2dict(testMessages.getValidBid()))
 
     def test_res2dict(self):
-        self.assertEqual(testMessages.validResDict, ipfs_channel.res2dict(testMessages.getValidRes()))
+        self.assertEqual(testMessages.validResDict, res2dict(testMessages.getValidRes()))
 
     def infochan_published_ask_handler(self, ask):
         self.assertEqual(testMessages.getValidAsk(), ask)

--- a/robonomics_liability/tests/test_signer.py
+++ b/robonomics_liability/tests/test_signer.py
@@ -5,7 +5,6 @@ from robonomics_msgs.msg import Result, Offer, Demand
 from robonomics_msgs import robonomicsMessageUtils
 from ethereum_common import eth_keyfile_helper
 from ethereum_common.msg import UInt256
-from robonomics_msgs import messageValidator
 import testMessages
 from binascii import hexlify
 
@@ -31,19 +30,19 @@ class TestSigner(unittest.TestCase):
 
     def test_demand_hash(self):
         test_nonce_value = UInt256("0")
-        rospy.logwarn("test_demand_hash: %s", hexlify(robonomicsMessageUtils.demand_hash(messageValidator.dict2ask(testMessages.validAskDict), nonce=test_nonce_value)))
+        rospy.logwarn("test_demand_hash: %s", hexlify(robonomicsMessageUtils.demand_hash(robonomicsMessageUtils.dict2demand(testMessages.validAskDict), nonce=test_nonce_value)))
         self.assertEqual(bytearray.fromhex('50f82ac8c4905d14a65632915fa21cf3726c720fd9650d3b5ac134ba2ebb9da2'),
-                         robonomicsMessageUtils.demand_hash(messageValidator.dict2ask(testMessages.validAskDict), nonce=test_nonce_value))
+                         robonomicsMessageUtils.demand_hash(robonomicsMessageUtils.dict2demand(testMessages.validAskDict), nonce=test_nonce_value))
 
     def test_offer_hash(self):
         test_nonce_value = UInt256("0")
-        rospy.logwarn("test_offer_hash: %s", hexlify(robonomicsMessageUtils.offer_hash(messageValidator.dict2bid(testMessages.validBidDict), nonce=test_nonce_value)))
+        rospy.logwarn("test_offer_hash: %s", hexlify(robonomicsMessageUtils.offer_hash(robonomicsMessageUtils.dict2offer(testMessages.validBidDict), nonce=test_nonce_value)))
         self.assertEqual(bytearray.fromhex('50f82ac8c4905d14a65632915fa21cf3726c720fd9650d3b5ac134ba2ebb9da2'),
-                         robonomicsMessageUtils.offer_hash(messageValidator.dict2bid(testMessages.validBidDict), nonce=test_nonce_value))
+                         robonomicsMessageUtils.offer_hash(robonomicsMessageUtils.dict2offer(testMessages.validBidDict), nonce=test_nonce_value))
 
     def test_result_hash(self):
         self.assertEqual(bytearray.fromhex('55c1a3f47762a8d67ed7c76bc140f85bdbf6eb41e70cc13aa41ed1791d939464'),
-                         robonomicsMessageUtils.result_hash(messageValidator.dict2res(testMessages.validResDict)))
+                         robonomicsMessageUtils.result_hash(robonomicsMessageUtils.dict2res(testMessages.validResDict)))
 
     def signed_ask_handler(self, ask):
         self.assertEqual(testMessages.validAskDict['model'], ask.model.multihash)

--- a/robonomics_msgs/src/robonomics_msgs/messageValidator.py
+++ b/robonomics_msgs/src/robonomics_msgs/messageValidator.py
@@ -3,18 +3,23 @@
 # Robonomics Demand/Offer/Result ipfs message converter
 #
 
+from .robonomicsMessageUtils import offer2dict, demand2dict, res2dict, dict2offer, dict2demand, dict2res
 from robonomics_msgs.msg import Demand, Offer, Result
-from ethereum_common.msg import Address, UInt256
-from ipfs_common.msg import Multihash
-from binascii import unhexlify
 import voluptuous as v
 import rospy
 
 
 @v.message('wrong hexadecimal field value')
-def isHexIntNotZero(arg):
-    if not int(arg, 16) > 0:
+def isHexIntNotZeroCanBeEmpty(arg):
+    if arg and not int(arg, 16) > 0:
         raise v.Invalid('value is zero')
+    return arg
+
+
+@v.message('wrong decimal field value')
+def isPositiveDecIntCanBeEmpty(arg):
+    if arg and not arg >= 0:
+        raise v.Invalid('value is negative')
     return arg
 
 
@@ -46,8 +51,8 @@ schemaAskBid = v.All(
         v.Required('cost'): v.All(int),
         v.Required('deadline'): v.All(int),
         v.Required('sender'): isAddress,
-        v.Required('nonce'): v.All(int),
-        v.Required('signature'): isHexIntNotZero()
+        v.Required('nonce'): isPositiveDecIntCanBeEmpty(),
+        v.Required('signature'): isHexIntNotZeroCanBeEmpty()
     })
 )
 
@@ -55,7 +60,7 @@ schemaResult = v.Schema({
     v.Required('liability'): isAddress,
     v.Required('result'): isIpfsBase58Hash,
     v.Required('success'): v.All(bool),
-    v.Required('signature'): isHexIntNotZero()
+    v.Required('signature'): isHexIntNotZeroCanBeEmpty()
 })
 
 schemaAskBidResult = v.Any(
@@ -64,95 +69,37 @@ schemaAskBidResult = v.Any(
 )
 
 
-def dict2ask(m):
-    msg = Demand()
-
-    msg.model = Multihash()
-    msg.model.multihash = m['model']
-
-    msg.objective = Multihash()
-    msg.objective.multihash = m['objective']
-
-    msg.token = Address()
-    msg.token.address = m['token']
-
-    msg.cost = UInt256()
-    msg.cost.uint256 = str(m['cost'])
-
-    msg.lighthouse = Address()
-    msg.lighthouse.address = m['lighthouse']
-
-    msg.validator = Address()
-    msg.validator.address = m['validator']
-
-    msg.validatorFee = UInt256()
-    msg.validatorFee.uint256 = str(m['validatorFee'])
-
-    msg.deadline = UInt256()
-    msg.deadline.uint256 = str(m['deadline'])
-
-    msg.sender = Address()
-    msg.sender.address = m['sender']
-
-    msg.nonce = UInt256()
-    msg.nonce.uint256 = str(m['nonce'])
-
-    msg.signature = unhexlify(m['signature'].encode('utf-8'))
-    return msg
+def isDemandFieldsCorrect(msg):
+    try:
+        assert isinstance(msg, Demand)
+        schemaAskBid(demand2dict(msg))
+        return True
+    except Exception as e:
+        rospy.logerr("Fields of demand message %s is not correct. Exception is %s", msg, e)
+        return False
 
 
-def dict2bid(m):
-    msg = Offer()
-
-    msg.model = Multihash()
-    msg.model.multihash = m['model']
-
-    msg.objective = Multihash()
-    msg.objective.multihash = m['objective']
-
-    msg.token = Address()
-    msg.token.address = m['token']
-
-    msg.cost = UInt256()
-    msg.cost.uint256 = str(m['cost'])
-
-    msg.validator = Address()
-    msg.validator.address = m['validator']
-
-    msg.lighthouse = Address()
-    msg.lighthouse.address = m['lighthouse']
-
-    msg.lighthouseFee = UInt256()
-    msg.lighthouseFee.uint256 = str(m['lighthouseFee'])
-
-    msg.deadline = UInt256()
-    msg.deadline.uint256 = str(m['deadline'])
-
-    msg.sender = Address()
-    msg.sender.address = m['sender']
-
-    msg.nonce = UInt256()
-    msg.nonce.uint256 = str(m['nonce'])
-
-    msg.signature = unhexlify(m['signature'].encode('utf-8'))
-    return msg
+def isOfferFieldsCorrect(msg):
+    try:
+        assert isinstance(msg, Offer)
+        schemaAskBid(offer2dict(msg))
+        return True
+    except Exception as e:
+        rospy.logerr("Fields of offer message %s is not correct. Exception is %s", msg, e)
+        return False
 
 
-def dict2res(m):
-    msg = Result()
-
-    msg.liability = Address()
-    msg.liability.address = m['liability']
-
-    msg.result = Multihash()
-    msg.result.multihash = m['result']
-
-    msg.success = m['success']
-    msg.signature = unhexlify(m['signature'].encode('utf-8'))
-    return msg
+def isResultFieldsCorrect(msg):
+    try:
+        assert isinstance(msg, Result)
+        schemaResult(res2dict(msg))
+        return True
+    except Exception as e:
+        rospy.logerr("Fields of result message %s is not correct. Exception is %s", msg, e)
+        return False
 
 
-def validateForAskBidResultBySchema(abr_msg):
+def msg_is_offer_demand_result(abr_msg):
     try:
         return schemaAskBidResult(abr_msg)
     except v.MultipleInvalid as e:
@@ -161,14 +108,14 @@ def validateForAskBidResultBySchema(abr_msg):
 
 
 def convertMessage(ipfsMessage):
-    validatedBySchema = validateForAskBidResultBySchema(ipfsMessage)
+    validatedBySchema = msg_is_offer_demand_result(ipfsMessage)
     if not (validatedBySchema is None):
         if 'validatorFee' in validatedBySchema:
             # rospy.logwarn('DEBUG: Message %s is valid Ask message', ipfsMessage)
-            return dict2ask(validatedBySchema)
+            return dict2demand(validatedBySchema)
         elif 'lighthouseFee' in validatedBySchema:
             # rospy.logwarn('DEBUG: Message %s is valid Bid ipfs message', ipfsMessage)
-            return dict2bid(validatedBySchema)
+            return dict2offer(validatedBySchema)
         else:
             # rospy.logwarn('DEBUG: Message %s is valid Result ipfs message', ipfsMessage)
             return dict2res(validatedBySchema)

--- a/robonomics_msgs/src/robonomics_msgs/robonomicsMessageUtils.py
+++ b/robonomics_msgs/src/robonomics_msgs/robonomicsMessageUtils.py
@@ -10,6 +10,9 @@ import multihash
 from web3.utils.normalizers import (
     abi_ens_resolver
 )
+from binascii import hexlify, unhexlify
+from ethereum_common.msg import Address, UInt256
+from ipfs_common.msg import Multihash
 
 
 def resolve_ens_name_to_address(name_address, web3):
@@ -93,3 +96,132 @@ def get_result_msg_sender_address(result):
     signature = result.signature
     recovered_hash = w3.eth.account.recoverHash(message_hash, signature=signature)
     return recovered_hash
+
+
+def offer2dict(b):
+    return {
+        'model': b.model.multihash,
+        'objective': b.objective.multihash,
+        'token': b.token.address,
+        'cost': int(b.cost.uint256),
+        'validator': b.validator.address,
+        'lighthouse': b.lighthouse.address,
+        'lighthouseFee': int(b.lighthouseFee.uint256),
+        'deadline': int(b.deadline.uint256),
+        'sender': b.sender.address,
+        'nonce': int(b.nonce.uint256) if b.nonce.uint256 else '',
+        'signature': hexlify(b.signature).decode('utf-8')
+    }
+
+
+def demand2dict(a):
+    return {
+        'model': a.model.multihash,
+        'objective': a.objective.multihash,
+        'token': a.token.address,
+        'cost': int(a.cost.uint256),
+        'lighthouse': a.lighthouse.address,
+        'validator': a.validator.address,
+        'validatorFee': int(a.validatorFee.uint256),
+        'deadline': int(a.deadline.uint256),
+        'sender': a.sender.address,
+        'nonce': int(a.nonce.uint256) if a.nonce.uint256 else '',
+        'signature': hexlify(a.signature).decode('utf-8')
+    }
+
+
+def res2dict(r):
+    return {
+        'liability': r.liability.address,
+        'result': r.result.multihash,
+        'success': r.success,
+        'signature': hexlify(r.signature).decode('utf-8')
+    }
+
+
+def dict2demand(m):
+    msg = Demand()
+
+    msg.model = Multihash()
+    msg.model.multihash = m['model']
+
+    msg.objective = Multihash()
+    msg.objective.multihash = m['objective']
+
+    msg.token = Address()
+    msg.token.address = m['token']
+
+    msg.cost = UInt256()
+    msg.cost.uint256 = str(m['cost'])
+
+    msg.lighthouse = Address()
+    msg.lighthouse.address = m['lighthouse']
+
+    msg.validator = Address()
+    msg.validator.address = m['validator']
+
+    msg.validatorFee = UInt256()
+    msg.validatorFee.uint256 = str(m['validatorFee'])
+
+    msg.deadline = UInt256()
+    msg.deadline.uint256 = str(m['deadline'])
+
+    msg.sender = Address()
+    msg.sender.address = m['sender']
+
+    msg.nonce = UInt256()
+    msg.nonce.uint256 = str(m['nonce'])
+
+    msg.signature = unhexlify(m['signature'].encode('utf-8'))
+    return msg
+
+
+def dict2offer(m):
+    msg = Offer()
+
+    msg.model = Multihash()
+    msg.model.multihash = m['model']
+
+    msg.objective = Multihash()
+    msg.objective.multihash = m['objective']
+
+    msg.token = Address()
+    msg.token.address = m['token']
+
+    msg.cost = UInt256()
+    msg.cost.uint256 = str(m['cost'])
+
+    msg.validator = Address()
+    msg.validator.address = m['validator']
+
+    msg.lighthouse = Address()
+    msg.lighthouse.address = m['lighthouse']
+
+    msg.lighthouseFee = UInt256()
+    msg.lighthouseFee.uint256 = str(m['lighthouseFee'])
+
+    msg.deadline = UInt256()
+    msg.deadline.uint256 = str(m['deadline'])
+
+    msg.sender = Address()
+    msg.sender.address = m['sender']
+
+    msg.nonce = UInt256()
+    msg.nonce.uint256 = str(m['nonce'])
+
+    msg.signature = unhexlify(m['signature'].encode('utf-8'))
+    return msg
+
+
+def dict2res(m):
+    msg = Result()
+
+    msg.liability = Address()
+    msg.liability.address = m['liability']
+
+    msg.result = Multihash()
+    msg.result.multihash = m['result']
+
+    msg.success = m['success']
+    msg.signature = unhexlify(m['signature'].encode('utf-8'))
+    return msg

